scripts/openai_server_demo/openai_api_server.py

```diff
@@ -0,0 +1,382 @@
+TEMPLATE_WITH_SYSTEM_PROMPT = (
+    "[INST] <sys> {system_prompt} </sys>\n" "{instruction} [/INST]"
+)
+
+TEMPLATE_WITHOUT_SYSTEM_PROMPT = "[INST] {instruction} [/INST]"
+
+def generate_prompt(
+    instruction, response="", with_system_prompt=False, system_prompt=None
+):
+    if with_system_prompt is True and system_prompt is not None:
+        prompt = TEMPLATE_WITH_SYSTEM_PROMPT.format_map(
+            {"instruction": instruction, "system_prompt": system_prompt}
+        )
+    else:
+        prompt = TEMPLATE_WITHOUT_SYSTEM_PROMPT.format_map({"instruction": instruction})
+    if len(response) > 0:
+        prompt += " " + response
+    return prompt
+
+
+def generate_completion_prompt(instruction: str):
+    """Generate prompt for completion"""
+    return generate_prompt(instruction, response="", with_system_prompt=False)
+
+
+def generate_chat_prompt(messages: list):
+    """Generate prompt for chat completion"""
+
+    system_msg = None
+    for msg in messages:
+        if msg.role == "system":
+            system_msg = msg.content
+    prompt = ""
+    is_first_user_content = True
+    for msg in messages:
+        if msg.role == "system":
+            continue
+        if msg.role == "user":
+            if is_first_user_content is True:
+                prompt += generate_prompt(
+                    msg.content, with_system_prompt=False, system_prompt=system_msg
+                )
+                is_first_user_content = False
+            else:
+                prompt += "<s>" + generate_prompt(msg.content, with_system_prompt=False)
+        if msg.role == "assistant":
+            prompt += f" {msg.content}" + "</s>"
+    return prompt
+
+
+def predict(
+    input,
+    max_new_tokens=512,
+    top_p=0.9,
+    temperature=0.2,
+    top_k=40,
+    num_beams=1,
+    repetition_penalty=1.1,
+    do_sample=True,
+    **kwargs,
+):
+    """
+    Main inference method
+    type(input) == str -> /v1/completions
+    type(input) == list -> /v1/chat/completions
+    """
+    if isinstance(input, str):
+        prompt = generate_completion_prompt(input)
+    else:
+        prompt = generate_chat_prompt(input)
+    inputs = tokenizer(prompt, return_tensors="pt")
+    input_ids = inputs["input_ids"].to(device)
+    attention_mask = inputs['attention_mask'].to(device)
+    generation_config = GenerationConfig(
+        temperature=temperature,
+        top_p=top_p,
+        top_k=top_k,
+        num_beams=num_beams,
+        do_sample=do_sample,
+        **kwargs,
+    )
+    generation_config.return_dict_in_generate = True
+    generation_config.output_scores = False
+    generation_config.max_new_tokens = max_new_tokens
+    generation_config.repetition_penalty = float(repetition_penalty)
+    with torch.no_grad():
+        generation_output = model.generate(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            eos_token_id=tokenizer.eos_token_id,
+            pad_token_id=tokenizer.pad_token_id,
```

Comment:
Using the tokenizer of chinese-mixtral may cause an error. It is suggested to change it to `pad_token_id=tokenizer.eos_token_id`.